aten/src/ATen/native/kleidiai/kai_kernels.cpp

```diff
@@ -0,0 +1,336 @@
+        kai_kernel_id::
+            matmul_clamp_f32_qsi8d32p1x8_qsi4c32p4x8_1x4x32_neon_dotprod);
+    return kai_pack_rhs_groupwise_int4<kai_matmul_ukernel_f32_qa8d32p_qs4c32p>(
+        kernel_packet, weight_packed, weight, bias, n, k, bl);
+  } else {
+    // Channelwise
+    auto kernel_packet = kai_select_channelwise_matmul_ukernel(
+        kai_kernel_id::
+            matmul_clamp_f32_qai8dxp1x8_qsi4cxp8x8_1x8x32_neon_dotprod);
+    return kai_pack_rhs_channelwise_int4<kai_matmul_ukernel_f32_qa8dxp_qs4cxp>(
+        kernel_packet, weight_packed, weight, scales, bias, n, k);
+  }
+}
+
+size_t kai_pack_rhs_int4_size(
+    const int64_t n,
+    const int64_t k,
+    const int64_t bl) {
+  if (32 == bl) {
+    // Groupwise 32
+    auto kernel_packet = kai_select_groupwise_matmul_ukernel(
+        kai_kernel_id::
+            matmul_clamp_f32_qsi8d32p1x8_qsi4c32p4x8_1x4x32_neon_dotprod);
+    const auto& ukernel = kernel_packet.ukernel;
+    const size_t nr = ukernel.get_nr();
+    const size_t kr = ukernel.get_kr();
+    return kernel_packet.kai_get_rhs_packed_size(n, k, nr, kr, bl);
+  } else {
+    // Channelwise
+    auto kernel_packet = kai_select_channelwise_matmul_ukernel(
+        kai_kernel_id::
+            matmul_clamp_f32_qai8dxp1x8_qsi4cxp8x8_1x8x32_neon_dotprod);
+    const auto& ukernel = kernel_packet.ukernel;
+    const size_t nr = ukernel.get_nr();
+    const size_t kr = ukernel.get_kr();
+    const size_t sr = ukernel.get_sr();
+    return kernel_packet.kai_get_rhs_packed_size(n, k, nr, kr, sr);
+  }
+}
+
+static void matmul_channelwise(
+    kai_matmul_ukernel_f32_qa8dxp_qs4cxp& kernel_packet,
+    size_t m_increment,
+    size_t m_start,
+    size_t m_per_thread,
+    size_t n_start,
+    size_t n_per_thread,
+    size_t n,
+    size_t k,
+    size_t mr,
+    size_t nr,
+    size_t kr,
+    size_t sr,
+    uint8_t* lhs_native_mtx_f32,
+    uint8_t* lhs_packed_mtx_qa8dx,
+    uint8_t* rhs_packed_mtx_qs4cx,
+    uint8_t* dst_act_mtx_f32) {
+  const size_t dst_stride = n * sizeof(float);
+  for (size_t m0 = 0; m0 < m_per_thread; m0 += m_increment) {
+    const float* src_ptr =
+        (const float*)(lhs_native_mtx_f32 +
+                       kai_get_lhs_offset_lhs_quant_pack_qai8dxp_f32(
+                           m_start + m0, k * sizeof(float)));
+    void* lhs_packed_ptr =
+        (void*)(lhs_packed_mtx_qa8dx +
+                kai_get_lhs_packed_offset_lhs_quant_pack_qai8dxp_f32(
+                    0, k, mr, kr, sr));
+    const void* rhs_packed_ptr =
+        (const void*)((const char*)rhs_packed_mtx_qs4cx +
+                      kernel_packet.ukernel.get_rhs_packed_offset(n_start, k));
+    float* dst_ptr = (float*)((uint8_t*)dst_act_mtx_f32 +
+                              kernel_packet.ukernel.get_dst_offset(
+                                  m_start + m0, n_start, dst_stride));
+
+    // Quantize and pack the Input
+    kernel_packet.kai_run_lhs_quant_pack(
```

Comment:
Is it OK to do dynamic quantization of activation for a weight only quant schema/op from accuracy point of view? i.e. why not make it explicit.

Reply:
The whole idea behind weight only quant is:
1. to quantize weights ahead of time and quantize input during runtime.
2. Leverage hardware int8 inrinsics to speed up matmul . [ (Input { int8 }) ⓧ  (Weight { int4->int8 } ]

Actually, doing dynamic quantisation on the input gives better accuracy as scales and zp are calculated during runtime and are tailored for that particular input as compared to static quantization which pre calculates scales, zp on calibration dataset before hand

Reply:
What do you mean when you say make it explicit?  Our Kernels literally say "quant_pack_lhs_int4_mm" in their name and we have a comment explaining same in the kai_kernels.h file.
I am open to suggestions

Reply:
So for weight only quant the compute happens in FP32. I am not arguing for static quant. 

> What do you mean when you say make it explicit? 

Apologies, I should have been more clear. I haven't seen the TorchAO PR for Kleidi yet but what I meant by more explicit is based on what I see in this PR is, if you see the quantization [flavors](https://github.com/pytorch/ao/blob/e283743b3cc4612bb641b88dca3670231724d396/torchao/quantization/quant_api.py#L427-L430), these Kleidi kernels aligns better with `int8_dynamic_activation_int4_weight` as opposed to `int4_weight_only`. The latter uses `torch.ops.aten._weight_int4pack_mm` which you are overloading here.

Reply:
I understand what you mean now. The initial version of this PR had separate operators for our kernels and did not use 
`torch.ops.aten._weight_int4pack_mm` for multiple reasons along with the one you mentioned. After discussion with Meta engineers, we decided to use _weight_int4pack_mm as the op interface for our kernel

Reply:
Do we already have a matmul op that uses` int8_dynamic_activation_int4_weight` in torchao or pytorch